framework/src/araui/assets/ui_font_asset.c

```diff
@@ -0,0 +1,297 @@
+	if (!body) {
+		UI_LOGE("error: out of memory!\n");
+		return UI_NULL;
+	}
+
+	memset(body, 0, sizeof(ui_font_asset_body_t));
+	((ui_asset_body_t *)body)->type = UI_FONT_ASSET;
+
+	font_data = (ui_font_data_t *)buf;
+	glyph_num = font_data->glyph_count;
+	body->height = font_data->font_height;
+	body->from_buf = true;
+
+	glyph_data = (ui_font_glyph_data_t *)(buf + font_data->header_size);
+
+	// read glyph from the buffer
+	if (ui_map_init(&body->glyph_map, DEFAULT_GLYPH_MAP_CAPACITY) != UI_OK) {
+		UI_LOGE("error: Cannot init the glyph map!\n");
+		UI_FREE(body);
+		return UI_NULL;
+	}
+
+	for (glyph_idx = 0; glyph_idx < glyph_num; glyph_idx++) {
+		glyph = (ui_asset_glyph_t *)UI_ALLOC(sizeof(ui_asset_glyph_t));
+		if (!glyph) {
+			_ui_font_asset_destroy_func((void *)body);
+			UI_LOGE("error: out of memory!\n");
+			return UI_NULL;
+		}
+
+		glyph->data = (uint8_t *)glyph_data + glyph_data->header_size;
+		glyph->size = glyph_data->data_size;
+		glyph->width = glyph_data->font_width;
+
+		if (ui_map_push(&body->glyph_map, glyph_data->utf8_code, glyph) != UI_OK) {
+			UI_LOGW("warning: 0x%08F is not added to the map!\n", utf8_code);
+		}
+
+		glyph_data = (ui_font_glyph_data_t *)((uint8_t *)glyph_data + glyph_data->header_size + glyph_data->data_size);
+	}
+
+	return (ui_asset_t)body;
+}
+
+ui_error_t ui_font_asset_destroy(ui_asset_t font)
+{
+	if (!ui_is_running()) {
+		return UI_NOT_RUNNING;
+	}
+
+	if (!font) {
+		return UI_INVALID_PARAM;
+	}
+
+	if (ui_request_callback(_ui_font_asset_destroy_func, (void *)font) != UI_OK) {
+		return UI_OPERATION_FAIL;
+	}
+
+	UI_LOGD("UI_OK.\n");
+
+	return UI_OK;
+}
+
+static void _ui_font_asset_destroy_func(void *userdata)
+{
+	size_t i;
+	ui_map_t *map;
+	ui_font_asset_body_t *body;
+	ui_list_t *list;
+	ui_list_node_t *node;
+	key_value_pair_t *pair;
+	ui_asset_glyph_t *glyph;
+
+	body = (ui_font_asset_body_t *)userdata;
+
+	// clear all data if this asset is not from buffer.
+	if (!body->from_buf) {
+		map = &body->glyph_map;
+		for (i = 0; i < map->hash_capacity; i++) {
+			list = &map->data[i];
+
+			node = ui_list_loop_begin(list);
+			while (node != ui_list_loop_end(list)) {
+				pair = (key_value_pair_t *)node->data;
+				glyph = (ui_asset_glyph_t *)pair->value;
+
+				UI_FREE(glyph->data);
+				UI_FREE(glyph);
+
+				node = node->next;
+			}
+		}
+	}
+
+	if (ui_map_deinit(&body->glyph_map) != UI_OK) {
+		UI_LOGE("error: Error on ui_map_deinit!\n");
+	}
+
+	UI_FREE(body);
+}
+
+size_t ui_font_asset_measure_text(ui_asset_t font, const char *text)
+{
+	size_t i;
+	size_t ret = 0;
+	size_t text_len;
+	uint32_t utf8_code;
+	uint8_t lobyte;
+	uint8_t hibyte;
+	ui_asset_glyph_t *glyph;
+
+	if (!ui_is_running()) {
+		UI_LOGE("error: UI framework is not running!\n");
+		return ret;
+	}
+
+	if (!font || !text) {
+		UI_LOGE("error: invalid parameter!\n");
+		return ret;
+	}
+
+	text_len = strlen(text);
+	for (i = 0; i < text_len; i++) {
+		utf8_code = 0;
+		if ((text[i] & 0x80) != 0x80) {
```

Comment:
In this case, the number is more readable. Because this number means binary values for bitwise arithmetic.

Reply:
If any of these value can be changed, define is better otherwise, you'd better put some comments to help prevent memory leak.

Reply:
These numbers are for bitwise operations.
If these values changed as definitions, there will be more confusion.

Reply:
I also think that using definition is better because when this bit mask needs to be changed, you should modify several lines or files. However, with definition, you just revise the definition one time.

Reply:
For example, if the 3 bytes satisfy this bitmask ( 110xxxxx 10xxxxxx 10xxxxxx ), 
It means korean character(2-bytes unicode).
In this case, we have to extract the 'x' values.
000xxxxx 00xxxxxx 00xxxxxx

How should I make definitions for them?